### PR TITLE
chore: upgraded subxt version

### LINKED ISSUE #418: Instant seal panic
For the integration tests we used to have feature-gated code duplication to not wait for finalization in the instant-seal integration tests. Since polkadot 0.9.31 finally includes the fix we have been waiting for, we shouldn't require this anymore - we should be able to use finalized events. However, when we switch to the finalized events, there is a problem that occurs only and consistently in `test_automatic_rbf_succeeds`, that gives the error shown below, forcing us to go back to the non-finalizing workaround we were using previously. We should investigate and make an upstream report.

```
Nov 23 11:50:14.364  INFO sc_consensus_manual_seal::rpc: Instant Seal success: CreatedBlock { hash: 0x85a148e068d10319110d4ffdd5fdc6876cf2bf7f3282da63829ddb4f2e21e74c, aux: ImportedAux { header_only: false, clear_justification_requests: false, needs_justification: false, bad_justification: false, is_new_best: true } }    
Nov 23 11:50:14.386  WARN jsonrpsee_core::server::rpc_module: unsubscribe call `author_unwatchExtrinsic` subscription key=SubscriptionKey { conn_id: 0, sub_id: Num(4234576671114396) } not an active subscription
thread 'tokio-runtime-worker' panicked at 'failed to create block: Rpc(ClientError(Call(Custom(ErrorObject { code: ServerError(14000), message: "Consensus Error: Import failed: Potential long-range attack: block not in finalized chain.", data: None }))))', /home/sander/workspace/interlay/polkabtc-clients/runtime/src/rpc.rs:172:14

### LINKED ISSUE #475: Upgrade the subxt dependency
We are currently using `0.25.0` and the latest version is `0.29.0`.

### PR DESCRIPTION
**Pull request by nakul1010**
PR Description:

Closes #475
Closes #418 

### Changes Implemented:
- [x] Modified dependencies.
- [x] Upgraded subxt version without utilizing the `substrate-compat` feature, thereby preventing conflicts with sp dependencies.
- [x] Imported necessary utils from subxt to the client, ensuring compatibility with client substrate dependencies.
- [x] Update test cases 
- [x] Added custom implementation for `AccountId`, `PairSigner` and `MultiSignature` since these types are feature gated by `substrate-compat` feature in subxt and hence can't be imported from subxt. 

### Alternate Approach:
- Considered forking subxt and updating the dependencies based on the node. However, this approach would require maintaining the forked version, thus it was discarded.

### References to Issues Opened on Subxt:
- [Sp Arithmetic subxt version clash](https://github.com/paritytech/subxt/issues/1039)
- [Static types](https://github.com/paritytech/subxt/issues/1040)
- [Sp dependencies version clash](https://github.com/paritytech/subxt/issues/1051)

### CODE REVIEW

File: runtime/src/types.rs [hunk 1]
```diff
@@ -103,24 +103,28 @@ mod metadata_aliases {
...
 
     pub type InterBtcReplaceRequest = metadata::runtime_types::interbtc_primitives::replace::ReplaceRequest<
-        AccountId,
+        Static<AccountId>,
         BlockNumber,
         Balance,
-        CurrencyId,
+        Static<CurrencyId>,
     >;
     pub type VaultId = metadata::runtime_types::interbtc_primitives::VaultId<AccountId, CurrencyId>;
```
Comment (nakul1010) [inline runtime/src/types.rs:129]: @sander2 and @gregdhill can i make the 
` pub type VaultId = metadata::runtime_types::interbtc_primitives::VaultId<Static<AccountId>, Static<CurrencyId>>;
` this will resolve the rest of the todos. But my question is will this cause issues in any of the upstream dependency for this type.  This is required because of the new wrapping done to the types [here](https://github.com/interlay/interbtc-clients/blob/7d7a9fe2234fd8aa3bb659ffa17ea00151b4df28/runtime/src/lib.rs#L201) and [here](https://github.com/interlay/interbtc-clients/blob/7d7a9fe2234fd8aa3bb659ffa17ea00151b4df28/runtime/src/lib.rs#L205). 

Reference: [Issue raised on subxt](https://github.com/paritytech/subxt/issues/1039)

Comment (gregdhill) [reply runtime/src/types.rs:129]: Happy to go with either solution as long as the tests pass. We can do a final review once the pipeline is green.

File: runtime/src/rpc.rs [hunk 2]
```diff
@@ -243,39 +249,38 @@ impl InterBtcParachain {
...
-    async fn query_finalized_or_error<Address>(
-        &self,
-        address: Address,
-    ) -> Result<<Address::Target as DecodeWithMetadata>::Target, Error>
+    async fn query_finalized_or_error<Address>(&self, address: Address) -> Result<Address::Target, Error>
     where
         Address: StorageAddress<IsFetchable = Yes>,
     {
-        self.query_finalized(address).await?.ok_or(Error::StorageItemNotFound)
+        self.query_finalized(address).await?.ok_or(Error::BlockNotFound)
```
Comment (gregdhill) [inline runtime/src/rpc.rs:270]: Why did you change this?

File: runtime/src/rpc.rs [hunk 3]
```diff
@@ -243,39 +249,38 @@ impl InterBtcParachain {
...
+    async fn query_finalized<Address>(&self, address: Address) -> Result<Option<Address::Target>, Error>
     where
         Address: StorageAddress<IsFetchable = Yes>,
     {
         let hash = self.get_finalized_block_hash().await?;
-        Ok(self.api.storage().fetch(&address, hash).await?)
+        match hash {
+            Some(hash) => Ok(self.api.storage().at(hash).fetch(&address).await?),
+            None => Ok(self.api.storage().at_latest().await?.fetch(&address).await?),
+        }
```
Comment (gregdhill) [inline runtime/src/rpc.rs:263]: Maybe we should change `get_finalized_block_hash` (to not return an `Option`) so this isn't necessary

File: bitcoin/tests/electrs.rs [hunk 4]
```diff
@@ -176,8 +179,8 @@ async fn should_create_transactions() -> Result<(), Error> {
 #[serial]
 async fn should_bump_fee() -> Result<(), Error> {
     let bitcoin_light = new_bitcoin_light();
-    fund_wallet(&bitcoin_light).await?;
 
+    fund_wallet(&bitcoin_light).await?;
```
Comment (gregdhill) [inline bitcoin/tests/electrs.rs:183]: Revert

File: faucet/src/http.rs [hunk 5]
```diff
@@ -11,15 +11,15 @@ use parity_scale_codec::{Decode, Encode};
 use reqwest::Url;
 use runtime::{
     AccountId, CollateralBalancesPallet, CurrencyId, Error as RuntimeError, InterBtcParachain, RuntimeCurrencyInfo,
-    Ss58Codec, TryFromSymbol, VaultRegistryPallet, SS58_PREFIX,
```
Comment (gregdhill) [inline faucet/src/http.rs:14]: As comment above, keep re-export

File: runner/Cargo.toml [hunk 6]
```diff
@@ -24,7 +24,7 @@ signal-hook = "0.3.14"
 signal-hook-tokio = { version = "0.3.1", features = ["futures-v0_3"] }
 futures = "0.3.21"
 backoff = { version = "0.3.0", features = ["tokio"] }
-subxt = "0.25.0"
+subxt = { version = "0.29.0", default_features = false,  features = ["jsonrpsee-ws"] }
```
[the remark this answers is not in the document]
Comment (gregdhill) [reply runner/Cargo.toml:27]: Is `default_features = false` required?

File: runtime/Cargo.toml [hunk 7]
```diff
@@ -33,45 +33,50 @@ url = "2"
...
 sp-arithmetic = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.40" }
+sp-core = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.40", default-features = false }
 # needs to agree with subxt's version
 sp-keyring = { version = "7.0.0", default_features = false }
 sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.40" }
 frame-support = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.40" }
 
 # Subxt dependencies
-subxt = "0.25.0"
+subxt = { version = "0.29.0", default_features = false,  features = ["jsonrpsee-ws"] }
```
Comment (gregdhill) [inline runtime/Cargo.toml:50]: Is `default_features = false` required?

File: runtime/src/assets.rs [hunk 8]
```diff
@@ -157,7 +157,7 @@ impl TryFromSymbol for CurrencyId {
     fn from_lend_token_symbol(symbol: &str) -> Option<Self> {
         let uppercase_symbol = symbol.to_uppercase();
         // Does the first character match the lend token prefix?
-        if uppercase_symbol.as_str().chars().next() == Some(LEND_TOKEN_SYMBOL_PREFIX) {
+        if uppercase_symbol.as_str().starts_with(LEND_TOKEN_SYMBOL_PREFIX) {
```
Comment (gregdhill) [inline runtime/src/assets.rs:160]: This is probably fine

File: runtime/src/rpc.rs [hunk 9]
```diff
@@ -146,7 +148,9 @@ impl InterBtcParachain {
         };
 
         parachain_rpc.store_assets_metadata().await?;
+
         parachain_rpc.store_lend_tokens().await?;
+
```
Comment (gregdhill) [inline runtime/src/rpc.rs:153]: Revert unnecessary line changes in this file

File: runtime/src/rpc.rs [hunk 10]
```diff
@@ -721,16 +720,18 @@ impl InterBtcParachain {
...
     where
-        T: Decode + Send + 'static,
+        T: Decode + Send + 'static + DecodeWithMetadata,
         U: Decode + Send + 'static,
     {
-        let head = self.get_finalized_block_hash().await?;
-        let mut iter = self.api.storage().iter(key_addr, DEFAULT_PAGE_SIZE, head).await?;
+        if cfg!(feature = "testing-utils") {
+            return Ok(vec![]);
+        }
```
Comment (gregdhill) [inline runtime/src/rpc.rs:728]: Remove, change `get_finalized_block_hash` to return latest block if `feature = "testing-utils"`

File: vault/tests/vault_integration_tests.rs [hunk 11]
```diff
@@ -1199,7 +1205,10 @@ mod test_with_bitcoind {
 
             // setup the service to test including necessary auxiliary services
             let shutdown_tx = ShutdownSender::new();
+
+            //here listening service starts, add println here as well
             let service = join(
+                // here is the error becaz of static i guess
```
Comment (gregdhill) [inline vault/tests/vault_integration_tests.rs:1211]: Revert

File: vault/src/refund.rs [hunk 12]
```diff
@@ -30,7 +30,7 @@ pub async fn listen_for_refund_requests(
                 tracing::info!("Received refund request: {:?}", event);
 
                 if !process_refunds {
-                    tracing::info!("Not processing refund");
+                    log::info!("Not processing refund");
```
Comment (gregdhill) [inline vault/src/refund.rs:33]: Revert

File: vault/src/refund.rs [hunk 13]
```diff
@@ -40,7 +40,7 @@ pub async fn listen_for_refund_requests(
                 let parachain_rpc = parachain_rpc.clone();
                 // Spawn a new task so that we handle these events concurrently
                 spawn_cancelable(shutdown_tx.subscribe(), async move {
-                    tracing::info!("Executing refund #{:?}", event.refund_id);
+                    log::info!("Executing refund #{:?}", event.refund_id);
```
Comment (gregdhill) [inline vault/src/refund.rs:43]: Revert

File: vault/src/relay/mod.rs [hunk 14]
```diff
@@ -107,7 +107,7 @@ impl<B: Backing, I: Issuing> Runner<B, I> {
             match self.backing.get_block_header(height).await? {
                 Some(header) => return Ok(header),
                 None => {
-                    tracing::trace!("No block found at height {}, sleeping for {:?}", height, self.interval);
+                    tracing::info!("No block found at height {}, sleeping for {:?}", height, self.interval);
```
Comment (gregdhill) [inline vault/src/relay/mod.rs:110]: Revert

File: vault/src/replace.rs [hunk 15]
```diff
@@ -192,10 +192,10 @@ pub async fn listen_for_execute_replace(
         .on_event::<ExecuteReplaceEvent, _, _, _>(
             |event| async move {
                 if &event.new_vault_id.account_id == parachain_rpc.get_account_id() {
-                    tracing::info!("Received event: execute replace #{:?}", event.replace_id);
+                    log::info!("Received event: execute replace #{:?}", *event.replace_id);
```
Comment (gregdhill) [inline vault/src/replace.rs:195]: Revert

**Review by gregdhill — CHANGES_REQUESTED** [review]
I've only reviewed about half of the files but I stopped because there are a lot of unnecessary changes. Please go through the diff yourself and revert anything unrelated to the subxt upgrade. In general please keep changes to a minimum so it is easier to review.

File: runner/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (nakul1010) [reply runner/Cargo.toml:27]: yes check it out [here](https://github.com/paritytech/subxt/blob/9a8fc33f2fe830009011e997773071099101d69b/subxt/Cargo.toml#L20)

File: runtime/Cargo.toml [hunk 16]
```diff
@@ -33,45 +33,50 @@ url = "2"
...
-interbtc = { package = "interbtc-parachain", git = "https://github.com/interlay/interbtc", rev = "7a2f5ee35d93feb0529a308fe87b42674bfca8e9", optional = true }
-interbtc-runtime = { package = "testnet-kintsugi-runtime-parachain", git = "https://github.com/interlay/interbtc", rev = "7a2f5ee35d93feb0529a308fe87b42674bfca8e9", optional = true }
+interbtc = { package = "interbtc-parachain", git = "https://github.com/interlay/interbtc", branch = "nakul/feat_add_trait", optional = true }
+interbtc-runtime = { package = "testnet-kintsugi-runtime-parachain", git = "https://github.com/interlay/interbtc", branch = "nakul/feat_add_trait", optional = true }
 rand = { version = "0.7", optional = true }
 
 [dependencies.primitives]
 git = "https://github.com/interlay/interbtc"
-rev = "7a2f5ee35d93feb0529a308fe87b42674bfca8e9"
+branch = "nakul/feat_add_trait"
```
[the remark this answers is not in the document]
Comment (nakul1010) [reply runtime/Cargo.toml:64]: Yes, there are changes on the node side, but there's a problem here `nakul/feat_add_trait` is branched out from commit id `7a2f5ee35d93feb0529a308fe87b42674bfca8e9` and not `master` since, the `master` doesn't contain testnet runtimes. The best solution is to merge PR #476 first, then `nakul/feat_add_trait` can be merged to master. After this the dependencies can point out to the master commit.

File: runtime/src/rpc.rs (hunk 2 above)
Comment (nakul1010) [reply runtime/src/rpc.rs:270]: reverting back

File: runtime/Cargo.toml (hunk 7 above)
Comment (nakul1010) [reply runtime/Cargo.toml:50]: mentioned in the [above comment](https://github.com/interlay/interbtc-clients/pull/483#discussion_r1259189848)

File: runtime/src/rpc.rs (hunk 3 above)
Comment (nakul1010) [reply runtime/src/rpc.rs]: The `get_finalized_block_hash` returns the latest block for the test feature.

File: runtime/src/tests.rs [hunk 17]
```diff
@@ -56,19 +59,29 @@ async fn test_getters() {
 
 #[tokio::test(flavor = "multi_thread")]
 async fn test_invalid_tx_matching() {
+    let bob_keyring = AccountKeyring::Alice;
```
Comment (gregdhill) [inline runtime/src/tests.rs:62]:
```suggestion
    let bob_keyring = AccountKeyring::Bob;
```

File: runtime/Cargo.toml (hunk 16 above)
[the remark this answers is not in the document]
Comment (gregdhill) [reply runtime/Cargo.toml:64]: Okay, please already open a PR to parachain master with the suggested changes.

File: vault/Cargo.toml [hunk 18]
```diff
@@ -53,6 +53,7 @@ faucet-rpc = { package = "faucet", path = "../faucet" }
 # Substrate dependencies
 # needs to agree with subxt's version
 sp-keyring = { version = "7.0.0" }
+subxt = { version = "0.29.0", default_features = false,  features = ["jsonrpsee-ws"] }
```
Comment (gregdhill) [inline vault/Cargo.toml:56]: Let's re-export from the runtime crate

File: vault/src/cancellation.rs [hunk 19]
```diff
@@ -245,7 +243,7 @@ impl<P: IssuePallet + ReplacePallet + UtilFuncs + SecurityPallet + Clone> Cancel
 
         match event {
             Event::ParachainBlock(height) => {
-                tracing::trace!(
+                tracing::info!(
```
Comment (gregdhill) [inline vault/src/cancellation.rs:246]: Revert

File: vault/src/cancellation.rs [hunk 20]
```diff
@@ -254,7 +252,7 @@ impl<P: IssuePallet + ReplacePallet + UtilFuncs + SecurityPallet + Clone> Cancel
                 Ok(self.cancel_requests::<T>(active_requests).await)
             }
             Event::BitcoinBlock(height) => {
-                tracing::trace!("Received Bitcoin block at height {} for {}", height, T::TYPE_NAME);
+                tracing::info!("Received Bitcoin block at height {} for {}", height, T::TYPE_NAME);
```
Comment (gregdhill) [inline vault/src/cancellation.rs:255]: Revert

File: vault/src/cancellation.rs [hunk 21]
```diff
@@ -391,10 +389,11 @@ mod tests {
         }
     }
 
+    use subxt::utils::Static;
     fn default_issue_request() -> InterBtcIssueRequest {
         InterBtcIssueRequest {
             amount: Default::default(),
-            btc_address: Default::default(),
+            btc_address: Static(Default::default()),
```
Comment (gregdhill) [inline vault/src/cancellation.rs:396]: Why is this required?

File: vault/src/delay.rs [hunk 22]
```diff
@@ -48,7 +48,7 @@ impl RandomDelay for OrderedVaultsDelay {
     /// * `data` - the seed used as a basis for the ordering (for example, an issue_id)
     async fn delay(&self, seed_data: &[u8; 32]) -> Result<(), RuntimeError> {
         fn hash_vault(data: &[u8; 32], account_id: &AccountId) -> sha256::Hash {
-            let account_id: [u8; 32] = account_id.clone().into();
+            let account_id: [u8; 32] = runtime::sp_core::crypto::AccountId32::from(account_id.0).into();
```
Comment (gregdhill) [inline vault/src/delay.rs:51]: Move import to top of file

File: runtime/src/rpc.rs [hunk 23]
```diff
@@ -1209,7 +1220,7 @@ impl SecurityPallet for InterBtcParachain {
     }
 
     /// Return any `ErrorCode`s set in the security module.
-    async fn get_error_codes(&self) -> Result<BTreeSet<ErrorCode>, Error> {
+    async fn get_error_codes(&self) -> Result<Vec<ErrorCode>, Error> {
```
Comment (gregdhill) [inline runtime/src/rpc.rs:1223]: Not sure why this changed but it doesn't matter since we will anyway remove these error codes in the next release

File: service/src/lib.rs [hunk 24]
```diff
@@ -179,7 +177,7 @@ where
             Ok(())
         }
         TerminationStatus::Completed(res) => {
-            tracing::trace!("Sending shutdown signal");
+            tracing::trace!("Received shutdown signal");
```
Comment (gregdhill) [inline service/src/lib.rs:180]: Revert

File: service/Cargo.toml [hunk 25]
```diff
@@ -9,6 +9,7 @@ async-trait = "0.1.40"
 futures = "0.3.5"
 clap = { version = "4.0.17", features = ["derive"]}
 thiserror = "1.0"
+log = "0.4.0"
```
Comment (gregdhill) [inline service/Cargo.toml:12]: Remove

File: runtime/src/types.rs [hunk 26]
```diff
@@ -180,7 +220,7 @@ mod account_id {
 
     impl PrettyPrint for AccountId {
         fn pretty_print(&self) -> String {
-            self.to_ss58check_with_version(SS58_PREFIX.into())
+            format!("{:?}", self.0)
```
Comment (gregdhill) [inline runtime/src/types.rs:223]: Revert

File: runtime/src/tests.rs [hunk 27]
```diff
@@ -178,16 +195,16 @@ async fn test_currency_id_parsing() {
...
+    // assert_eq!(CurrencyId::try_from_symbol("qkInt".to_string()).unwrap(), LendToken(1));
+    // // Even if matching as a qToken fails, the foreign asset should still be found.
+    // // Matching "QQQ" will recursively call `try_from_symbol` function three more times (including
+    // // with an empty string), because `Q` matches the lend token prefix each time.
+    // assert_eq!(CurrencyId::try_from_symbol("qQQ".to_string()).unwrap(), ForeignAsset(3));
+    // assert_eq!(
+    //     CurrencyId::try_from_symbol("TeSt".to_string()).unwrap(),
+    //     ForeignAsset(2)
+    // );
+    // assert_eq!(ForeignAsset(2).decimals().unwrap(), 10);
```
Comment (gregdhill) [inline runtime/src/tests.rs:209]: Revert

File: runtime/src/rpc.rs [hunk 28]
```diff
@@ -217,15 +219,17 @@ impl InterBtcParachain {
         Self::new(ws_client, signer, shutdown_tx).await
     }
 
-    async fn get_fresh_nonce(&self) -> u32 {
+    async fn get_fresh_nonce(&self) -> Result<u32, Error> {
```
Comment (gregdhill) [inline runtime/src/rpc.rs:222]: Not sure if we want this to return an error since it may bypass the existing error handling, worth checking

File: runtime/src/types.rs [hunk 29]
```diff
@@ -1,35 +1,35 @@
...
+// also multi address and what not :)
+pub type AccountId = utils_accountid::AccountId32;
+pub type MultiSignature = multi_signature::MultiSignature;
 pub type Balance = primitives::Balance;
 pub type Index = u32;
 pub type BlockNumber = u32;
-pub type H160 = subxt::ext::sp_core::H160;
-pub type H256 = subxt::ext::sp_core::H256;
-pub type U256 = subxt::ext::sp_core::U256;
+pub type H160 = sp_core::H160;
```
Comment (gregdhill) [inline runtime/src/types.rs:23]: Does subxt not still re-export sp-core?

File: runtime/src/tests.rs [hunk 30]
```diff
@@ -56,19 +59,29 @@ async fn test_getters() {
...
 
 #[tokio::test(flavor = "multi_thread")]
 async fn test_too_low_priority_matching() {
+    let alice_keyring = AccountKeyring::Alice;
+    let alice_substrate_account = alice_keyring.to_account_id();
+    let alice = AccountId32(alice_substrate_account.clone().into());
+
+    let bob_keyring = AccountKeyring::Alice;
+    let bob_substrate_account = bob_keyring.to_account_id();
+    let bob = AccountId32(bob_substrate_account.clone().into());
```
Comment (gregdhill) [inline runtime/src/tests.rs:80]:
```suggestion
    let bob_keyring = AccountKeyring::Bob;
    let bob_substrate_account = bob_keyring.to_account_id();
    let bob = AccountId32(bob_substrate_account.clone().into());
```

File: runtime/src/rpc.rs [hunk 31]
```diff
@@ -1629,10 +1650,13 @@ impl VaultRegistryPallet for InterBtcParachain {
...
     async fn get_vault(&self, vault_id: &VaultId) -> Result<InterBtcVault, Error> {
-        match self
+        let vault: InterBtcVault = self
             .query_finalized(metadata::storage().vault_registry().vaults(vault_id))
             .await?
-        {
+            .unwrap()
+            .into();
+        let vault = Some(vault);
+        match vault {
```
Comment (gregdhill) [inline runtime/src/rpc.rs:1659]: Why are you doing this?

File: runtime/src/rpc.rs [hunk 32]
```diff
@@ -1629,10 +1650,13 @@ impl VaultRegistryPallet for InterBtcParachain {
     /// * `VaultNotFound` - if the rpc returned a default value rather than the vault we want
     /// * `VaultLiquidated` - if the vault is liquidated
     async fn get_vault(&self, vault_id: &VaultId) -> Result<InterBtcVault, Error> {
-        match self
+        let vault: InterBtcVault = self
             .query_finalized(metadata::storage().vault_registry().vaults(vault_id))
             .await?
-        {
+            .unwrap()
```
Comment (gregdhill) [inline runtime/src/rpc.rs:1656]: No unwraps please

**Review by gregdhill — CHANGES_REQUESTED** [review]

File: runtime/src/types.rs (hunk 29 above)
Comment (nakul1010) [reply runtime/src/types.rs:23]: we can't use `subxt::ext:: sp_core` it's feature gated by `substrate-compat` feature flag

File: vault/src/cancellation.rs (hunk 21 above)
Comment (nakul1010) [reply vault/src/cancellation.rs:396]: `Address` enum in node primitives use  `sp_core::H160` & `sp_core::H256`, which don't implement `scale_encode` & `scale_decode` traits. Therefore static is needed another option discussed was to use `subxt::sp_core` which is feature gated by `substrate-compact` feature.

File: runtime/src/rpc.rs (hunk 28 above)
Comment (nakul1010) [reply runtime/src/rpc.rs:212]: IF we don't pass out an error back to the caller, we need to add a unwrap/error log.

File: runtime/src/rpc.rs (hunk 31 above)
Comment (nakul1010) [reply runtime/src/rpc.rs]: removed

File: runtime/src/rpc.rs [hunk 33]
```diff
@@ -320,12 +311,8 @@ impl InterBtcParachain {
         .await
     }
 
-    pub async fn get_finalized_block_hash(&self) -> Result<Option<H256>, Error> {
-        if cfg!(feature = "testing-utils") {
-            Ok(None)
-        } else {
-            Ok(Some(self.api.rpc().finalized_head().await?))
-        }
+    pub async fn get_finalized_block_hash(&self) -> Result<H256, Error> {
```
Comment (nakul1010) [inline runtime/src/rpc.rs:314]: No need for having `testing-utils` feature flag unable, as `instant-seal ` does finalize the block.

Comment (gregdhill) [reply runtime/src/rpc.rs:314]: Are you saying this is solved? https://github.com/interlay/interbtc-clients/issues/418

File: runtime/src/rpc.rs [hunk 34]
```diff
@@ -165,11 +164,10 @@ impl InterBtcParachain {
             pub aux: ImportedAux,
         }
 
-        let head = self.get_finalized_block_hash().await.unwrap();
```
[the remark this answers is not in the document]
Comment (gregdhill) [reply runtime/src/rpc.rs:168]: Blocks will be automatically finalized with instant-seal, but we previously ran into this error: https://github.com/interlay/interbtc-clients/issues/418

File: runtime/src/rpc.rs [hunk 35]
```diff
@@ -165,11 +164,10 @@ impl InterBtcParachain {
...
         let _: CreatedBlock<interbtc_runtime::Hash> = self
             .api
             .rpc()
-            .request("engine_createBlock", rpc_params![true, true, head])
+            .request("engine_createBlock", rpc_params![true, true, None::<H256>])
```
Comment (gregdhill) [inline runtime/src/rpc.rs:170]: Does this work?
```suggestion
            .request("engine_createBlock", rpc_params![true, true])
```

File: runtime/src/rpc.rs (hunk 28 above)
Comment (gregdhill) [reply runtime/src/rpc.rs:207]: What are the possible errors now incurred? I think in `notify_retry` we will match some specific errors if we will check specifically if we can retry or propagate the error.

File: runtime/src/rpc.rs (hunk 34 above)
[the remark this answers is not in the document]
Comment (nakul1010) [reply runtime/src/rpc.rs:168]: Yes, after removing the feature flag, I didn't face the error mentioned in the issue. I suspect the error is not arising from upstream dependencies but rather from block creation parameters [here](https://github.com/interlay/interbtc-clients/blob/eaf5838959e2c5d01058e85fa5c5e5808fbbbd78/runtime/src/rpc.rs#L172).

File: runtime/src/rpc.rs (hunk 28 above)
Comment (nakul1010) [reply runtime/src/rpc.rs:207]: It will pass out a subxt runtime error defined [here](https://github.com/interlay/interbtc-clients/blob/6b5933f01a4ef0b602a56a9acd31965b80dc89e7/runtime/src/error.rs#L81). I have moved the [get_fresh_nonce](https://github.com/interlay/interbtc-clients/blob/6b5933f01a4ef0b602a56a9acd31965b80dc89e7/runtime/src/rpc.rs#L268C34-L268C49) method inside  [match timeout](https://github.com/interlay/interbtc-clients/blob/6b5933f01a4ef0b602a56a9acd31965b80dc89e7/runtime/src/rpc.rs#L269) in latest commit.

And i don't think there is a need to update the retry/propagate checks as the [create_signed_with_nonce](https://github.com/interlay/interbtc-clients/blob/6b5933f01a4ef0b602a56a9acd31965b80dc89e7/runtime/src/rpc.rs#L273) method returns the same subxt error type, and has been already handle in the master branch so same checks should work for `get_fresh_nonce` method as well.

File: runtime/src/rpc.rs [hunk 36]
```diff
@@ -232,39 +232,30 @@ impl InterBtcParachain {
...
-        &self,
-        address: Address,
-    ) -> Result<<Address::Target as DecodeWithMetadata>::Target, Error>
+    async fn query_finalized_or_default<Address>(&self, address: Address) -> Result<Option<Address::Target>, Error>
     where
         Address: StorageAddress<IsFetchable = Yes, IsDefaultable = Yes>,
     {
         let hash = self.get_finalized_block_hash().await?;
-        Ok(self.api.storage().fetch_or_default(&address, hash).await?)
+        Ok(Some(self.api.storage().at(hash).fetch_or_default(&address).await?))
```
Comment (sander2) [inline runtime/src/rpc.rs:258]: Why wrap in option? Looks like you never return none?

File: runtime/src/types.rs [hunk 37]
```diff
@@ -173,10 +211,9 @@ pub trait PrettyPrint {
 
 mod account_id {
     use super::*;
-
     impl PrettyPrint for AccountId {
         fn pretty_print(&self) -> String {
-            self.to_ss58check_with_version(SS58_PREFIX.into())
+            self.to_ss58check()
```
Comment (gregdhill) [inline runtime/src/types.rs:216]: Can we keep this? It is useful to print the account ID encoded for the chain

File: runtime/src/utils/account_id.rs [hunk 38]
```diff
@@ -0,0 +1,177 @@
...
+        AccountId32(value)
+    }
+
+    // Return the ss58-check string for this key. Adapted from `sp_core::crypto`. We need this to
+    // serialize our account appropriately but otherwise don't care.
+    pub fn to_ss58check(&self) -> String {
+        // For serializing to a string to obtain the account nonce, we use the default substrate
+        // prefix (since we have no way to otherwise pick one). It doesn't really matter, since when
+        // it's deserialized back in system_accountNextIndex, we ignore this (so long as it's valid).
+        const SUBSTRATE_SS58_PREFIX: u8 = 42;
```
Comment (gregdhill) [inline runtime/src/utils/account_id.rs:55]: You could also use the proper SS58 prefix here since it is defined in this crate

File: vault/tests/vault_integration_tests.rs [hunk 39]
```diff
@@ -47,9 +50,16 @@ where
             .flat_map(|account_id| {
                 vec![DEFAULT_TESTING_CURRENCY, DEFAULT_NATIVE_CURRENCY]
                     .into_iter()
-                    .map(move |currency_id| (account_id.clone(), 1 << 60, 0, currency_id))
+                    .map(move |currency_id| {
+                        (
+                            runtime::utils_accountid::AccountId32::from(account_id.clone()),
```
Comment (gregdhill) [inline vault/tests/vault_integration_tests.rs:55]:
```suggestion
                            account_id.clone().into(),
```

File: runtime/src/rpc.rs (hunk 33 above)
Comment (gregdhill) [reply runtime/src/rpc.rs:314]: Looks like it is, @nakul1010 if this closes that issue please add a keyword to the PR description

File: runtime/src/rpc.rs (hunk 36 above)
Comment (nakul1010) [reply runtime/src/rpc.rs]: removed Option

File: runtime/src/utils/account_id.rs (hunk 38 above)
Comment (nakul1010) [reply runtime/src/utils/account_id.rs]: imported `SS58_PREFIX` when converting account id in `to_ss58check` method

File: runtime/src/utils/account_id.rs [hunk 40]
```diff
@@ -0,0 +1,177 @@
+use base58::{FromBase58, ToBase58};
+use blake2::{Blake2b512, Digest};
+use codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+use std::convert::TryInto;
+
+/// A 32-byte cryptographic identifier. This is a simplified version of Substrate's
```
[the remark this answers is not in the document]
Comment (nakul1010) [reply runtime/src/utils/account_id.rs:7]: Added 4th point in the description.

File: runtime/src/utils/account_id.rs [hunk 41]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    // Return the ss58-check string for this key. Adapted from `sp_core::crypto`. We need this to
+    // serialize our account appropriately but otherwise don't care.
+    pub fn to_ss58check(&self) -> String {
+        // For serializing to a string to obtain the account nonce, we use the default substrate
+        // prefix (since we have no way to otherwise pick one). It doesn't really matter, since when
+        // it's deserialized back in system_accountNextIndex, we ignore this (so long as it's valid).
+        // prefix <= 63 just take up one byte at the start:
+        let mut v = vec![SS58_PREFIX as u8];
```
Comment (sander2) [inline runtime/src/utils/account_id.rs:56]: Can you please add a test? It seems to me like this shouldn't work - `SS58_PREFIX` is either 2032 or 2092, both of which don't fit into u8. Please also add permalinks to where you copied this from.

See https://github.com/paritytech/substrate/blob/696d09b3d1da1b2045fe46200df1d9f6d6d9d829/primitives/core/src/crypto.rs#L328-L340 for how to  use prefixes > 256

Comment (nakul1010) [reply runtime/src/utils/account_id.rs]: I have referenced it from [subxt utils](https://github.com/paritytech/subxt/tree/master/subxt/src/utils). Do you want me to modify it according to `sp_core` reference provided above or depend on subxt itself?

File: runtime/src/utils/account_id.rs (hunk 40 above)
[the remark this answers is not in the document]
Comment (nakul1010) [reply runtime/src/utils/account_id.rs:8]: added doc comments in the file and opened issue #493 for more discussion regrading the use of custom types.

File: runtime/src/utils/account_id.rs (hunk 41 above)
Comment (nakul1010) [reply runtime/src/utils/account_id.rs]: Modified implementation according to `sp_core`

**Review by sander2 — APPROVED** [review]
